main.cpp

```diff
@@ -1261,33 +1293,36 @@ void push_message(byte type, uint32_t lval, float fval, const char* sval) {
         pd.read(lval, &prog);
         if(lval<pd.nprograms) strcat(postval, prog.name);
       }
-      strcat_P(postval, PSTR(" with "));
-      itoa((int)fval, postval+strlen(postval), 10);
-      strcat_P(postval, PSTR("% water level."));
-      break;
-
-    case IFTTT_RAINSENSOR:
+      sprintf_P(postval+strlen(postval), PSTR(" with %f%% water level."), fval);
+    }
+    break;
 
+  case NOTIFY_RAINSENSOR:
+    if (os.options[OPTION_MQTT_ENABLE]&test_type) {
+      strcpy_P(topic, PSTR("opensprinkler/sensor/rain"));
+      sprintf_P(payload, PSTR("{\"state\":%f}"), fval);
+    }
+    if (os.options[OPTION_IFTTT_ENABLE]&test_type) {
       strcat_P(postval, (lval==LOGDATA_RAINDELAY) ? PSTR("Rain delay ") : PSTR("Rain sensor "));
       strcat_P(postval, ((int)fval)?PSTR("activated."):PSTR("de-activated"));
+    }
+    break;
 
-      break;
-
-    case IFTTT_FLOWSENSOR:
-      strcat_P(postval, PSTR("Flow count: "));
-      itoa(lval, postval+strlen(postval), 10);
-      strcat_P(postval, PSTR(", volume: "));
-      {
-      uint32_t volume = os.options[OPTION_PULSE_RATE_1];
-      volume = (volume<<8)+os.options[OPTION_PULSE_RATE_0];
-      volume = lval*volume;
-      itoa(volume/100, postval+strlen(postval), 10);
-      strcat(postval, ".");
-      itoa(volume%100, postval+strlen(postval), 10);
-      }
-      break;
+  case NOTIFY_FLOWSENSOR:
+    volume = os.options[OPTION_PULSE_RATE_1];
+    volume = (volume<<8)+os.options[OPTION_PULSE_RATE_0];
+    volume = lval*volume;
+    if (os.options[OPTION_MQTT_ENABLE]&test_type) {
+      strcpy_P(topic, PSTR("opensprinkler/sensor/flow"));
+      sprintf_P(payload, PSTR("{\"count\":%d,\"volume\":%.2f}"), lval, (float)volume/100);
+    }
+    if (os.options[OPTION_IFTTT_ENABLE]&test_type) {
+      sprintf_P(postval+strlen(postval), PSTR("Flow count: %d, volume: %.2f"), lval, (float)volume/100);
```

Comment:
Are you sure that arduino `sprintf` supports floats ? I remember that you had jump through hoops when it came to formatting floats and use `dtostrf()`

Reply:
You're right https://stackoverflow.com/questions/27651012/arduino-sprintf-float-not-formatting